joomla/joomla-cms
# Fix findLabel() performance being bad if labels are missing and being O(n^n) when they don't have tag id like ID-lbl

#### Summary of Changes

If you have a lot of form elements e.g. buttons that do not have a label or if the form has labels that do not have HTML tag ID like *-lbl, then findLabel() performance does a full DOM search using selector:
var $label = $('label[for="'+$elem.attr('id')+'"]');

The above describe findLabel() behaviour of:
- validate.js

For:
- html5fallback.js
  the full DOM search is always performed
#### Testing Instructions

Apply the patch the edit file (or some other form file):
and replace:

```
</form>
```

with:

```
<?php
echo '
<script>
    var test_cnt = 0;

    function time_test_Isvalid(el)
    {
        window.console.log( \'CALLING isValid()\' );

        var vTimeStart = new Date();
        document.formvalidator.isValid(document.getElementById(el.form.id));
        var vTimeDiff = (new Date())  - vTimeStart;

        window.console.log( \'isValid() time: \' + vTimeDiff );

        return false;
    }

    function inject_test_input(el, box_id)
    {
        var box = document.getElementById(box_id);

        var lbl = document.createElement(\'label\');
        lbl.setAttribute(\'for\', \'testid_\' + test_cnt);
        lbl.setAttribute(\'class\', \'label\');
        lbl.innerHTML = \'New element\';
        box.appendChild(lbl);

        var input = document.createElement(\'input\');
        input.type = \'text\';
        input.setAttribute(\'class\', \'required\');
        input.id = \'testid_\' + test_cnt;
        box.appendChild(input);

        test_cnt++;
        window.console.log(\'Element injected\');
        return false;
    }   
</script>

<br/>
<button  onclick="time_test_Isvalid(this);  return false;"  class="btn">  Test validation </button>
<button  onclick="inject_test_input(this, \'inject_test_box\'); time_test_Isvalid(this); return false;"  class="btn">   Inject field + Test validation </button>
<div id="inject_test_box"></div>
';
$use_labels = 1;
$add_for = 1;
$add_dash_lbl = 1;
for($i = 0; $i < 10000; $i++)
{
    if ($use_labels)
    {
        $lbl_for = $add_for  ?  'for="dummy_field_'.$i.'"'  :  '';
        $lbl_id  = $add_dash_lbl  ?  'id="dummy_field_'.$i.'-lbl"'  :  'id="dummy_field_'.$i.'-nodashlbl"';
        echo '<label '.$lbl_for.' '.$lbl_id.' class="label">Element '.$i.'</label> ';
    }
    echo '<input id="dummy_field_'.$i.'" type="text" name="dummy_field_'.$i.'" /> ';
}
?>

</form>
```

The above will 
- add 10,000 form elements into the form
- add a button to trigger and TIME validation of isValid() (validate.js) without submitting the form
  (open console to see output)
- add a button to inject a new form element into the DOM and re-trigger validation, so that we can verify that validation behaviour and performance after validation has been trigger once and some element has been added after it

**Validations** (testing times incude only isValid() of validate.js):

```
All fields have labels with ID-lbl
             1st      2nd, +           After Injecting new field
BEFORE:      1070ms    about 250ms      250ms - 550ms
AFTER:        630ms    about 250ms      370ms - 630ms

All 10,000 labels are missing, e.g. input buttons or buttons
             1st      2nd, +           After Injecting new field
BEFORE:     5700ms   about 180ms    180ms - 430ms
AFTER:       450ms   about 180ms    230ms - 480ms


All 10,000 labels don't have id like ID-lbl
             1st       2nd, +         After Injecting new field
BEFORE:   28700ms    about 220ms      220ms - 440ms
AFTER:      620ms    about 220ms      360ms - 650ms
```
